Add dependent package lists to pkg.info

na4zagin3 opened the pull request:
This PR adds `packages` field to `pkg.info` files. The field has an object whose keys are target distributions and the values are dependent package names.
Having an empty array as a value shows there is no dependent packages.

Added target distributions are `fedora`, `ubuntu`, and `ubuntu-ci`. `ubuntu-ci` is for `GtkSheet` which is not available in typical Ubuntu distributions.

### Motivation

I feel a bit difficulty to add `IBus` to a line in `PKGS.sh` and `libibus-1.0-dev` to a line to `.github/workflows/linux.yml` because it won't scale when we have multiple PRs to add bindings.

I'd like to have pkg.info file dependencies to extract a package list and a dependency list from them.

### Notes / Other Considerations
#### Fedora Packages
I left package lists for `fedora` empty because I couldn't find a search engine that I can get package names from file names. I know Ubuntu provides such a search engine but I don't know about Fedora because I don't use Fedora.

#### Location of Package Information
BTW, adding to `pkg.info` may not be a good option. Should I add a new file for such information?

#### Field Name
There may be better field names than `packages`, e.g., `distributionPackages`.

#### Multiple Ubuntu Versions
`libhandy-0.0-dev` is available since Ubuntu 19.10. How should we handle multiple Ubuntu versions?

garetxe commented:
This is great, thanks for the work! I had been wanting to do something like this for a while. Some replies to your comments, and a question below:

> I left package lists for fedora empty

That's all right, I can fill this in

> Should I add a new file for such information?

No, I think putting this in `pkg.info` is a good choice.

> There may be better field names than packages, e.g., distributionPackages.

Yes, I think that I would prefer `distributionPackages`. Could you please modify the PR to use this name?

> libhandy-0.0-dev is available since Ubuntu 19.10. How should we handle multiple Ubuntu versions?

I wouldn't overcomplicate things here, making `ubuntu` mean "a new enough ubuntu version" is probably good enough.

One question/comment: did you try whether `cabal run genBuildInfo $(./PKGS.sh)` still works in the bindings directory? Either way the format of `pkg.info` and `ProjectInfo` in `bindings/ProjectInfo.hs` should be kept in sync.

na4zagin3 commented:
I'm glad to hear that!

> > I left package lists for fedora empty
> 
> That's all right, I can fill this in
>
> > Should I add a new file for such information?
> 
> No, I think putting this in `pkg.info` is a good choice.
>
> > libhandy-0.0-dev is available since Ubuntu 19.10. How should we handle multiple Ubuntu versions?
> 
> I wouldn't overcomplicate things here, making `ubuntu` mean "a new enough ubuntu version" is probably good enough.

Got it. Thanks!
 
> > There may be better field names than packages, e.g., distributionPackages.
> 
> Yes, I think that I would prefer `distributionPackages`. Could you please modify the PR to use this name?

Sure! I renamed the field name.

> One question/comment: did you try whether `cabal run genBuildInfo $(./PKGS.sh)` still works in the bindings directory? Either way the format of `pkg.info` and `ProjectInfo` in `bindings/ProjectInfo.hs` should be kept in sync.

The execution result of `./PKGS.sh` return the same set of packages with different order (i.e., alphabetical order in the current locale instead of the given order) and different delimiter (i.e., line breaks instead of spaces).  As long as genBuildInfo does not matter the order of the given packages, it should be fine.  `cabal new-run genBuildInfo $(./PKGS.sh)` worked when I tried enabling bindings for packages installed in my environment during.

I also updated `ProjectInfo.hs`.

garetxe commented:
Great, merged. Thanks a lot!
